client/src/cordova/apple/OutlineAppleLib/Sources/OutlineError/DetailedJsonError.swift

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Tun2socks
+
+/// Defines keys used in the userInfo of a NSError.
+private enum DetailedJsonErrorKeys {
+  static let Code: String = "DetailedJsonErrorCode"
+  static let ErrorJson: String  = "DetailedJsonErrorDetails"
+}
+
+/// An error type representing an error with a code and a JSON string containing detailed information.
+public class DetailedJsonError: Error {
+  private let code: String
+  private let json: String
+
+  /// Create a new DetailedJsonError with a specified error code string and details in JSON.
+  public init(withErrorCode code: String, andErrorJson json: String) {
+    self.code = code
+    self.json = json
+  }
+
+  /// Create a new DetailedJsonError from a Go's PlatformError object.
+  public convenience init(fromPlatformError perr: PlaterrorsPlatformError) {
+    var marshalErr: NSError?
+    var errorJson = PlaterrorsMarshalJSONString(perr, &marshalErr)
+    if marshalErr != nil {
+      errorJson = "error code = \(perr.code), failed to fetch details"
+    }
+    self.init(withErrorCode: perr.code, andErrorJson: errorJson)
+  }
+
+  /// Create a new DetailedJsonError with a specified error code string and message.
+  /// It will marshal the error code and message into detailed JSON.
+  public convenience init(withErrorCode code: String, andMessage message: String) {
+    guard let perr = PlaterrorsNewPlatformError(code, message) else {
+      self.init(withErrorCode: code, andErrorJson: "failed to marshal: \(message)")
+      return
+    }
+    self.init(fromPlatformError: perr)
+  }
+
+  /// Create a new DetailedJsonError from a NSError object.
+  /// It will look for the JSON details from the userInfo.
+  /// If no details found, it will use the localizedDescription as the message.
+  public convenience init(fromError err: Error, withErrorCode errCode: String? = nil) {
```

Comment:
Why so many constructors? I find that excessive overloading hursts readability. Much better to have a single constructor and factory functions.

Reply:
These are all `convenience` constructors to convert from other type of errors to `DetailedJsonError`. Making them to conversion functions sounds good to me.

Reply:
Converted to `from` function.